src/server/tran_server.cpp

```diff
@@ -156,14 +156,32 @@ tran_server::push_request (tran_to_page_request reqid, std::string &&payload)
 int
 tran_server::send_receive (tran_to_page_request reqid, std::string &&payload_in, std::string &payload_out)
 {
+  int err_code = NO_ERROR;
+  constexpr auto timeout_1m = std::chrono::minutes { 1 };  // 1m, TODO: should be configurable.
   std::shared_lock<std::shared_mutex> s_lock (m_page_server_conn_vec_mtx);
-  if (m_page_server_conn_vec.empty())
-    {
-      assert_release (false); // TODO some error-handling is needed such as shutdown
-      return ER_FAILED; // All connections have been disconnected already
-    }
+  /*
+   *  send_receive () on a connection could fail when it is disconnected while waiting for its response.
+   *  In this case, it waits and retries with the expectation that the main connection will be changed.
+   *  - Normal disconnection: in receive_disconnect_request() and disconnect_all_page_servers()
+   *  - Abnormal disconnection: in the error handler, here, or somewhere else (TODO)
+   */
 
-  return m_page_server_conn_vec[0]->send_receive (reqid, std::move (payload_in), payload_out);
+  // TODO: exits when a thread waiting on it (transaction or whatever) stops. It should wake up periodically and check it.
```

Comment:
Don't understand this `TODO`.

Reply:
One of what we discussed. When shutdown a server, any transaction or thread should have a way out of this wait().